rust/otap-dataflow/crates/otap/src/retry_processor.rs

```diff
@@ -290,52 +228,123 @@ impl Processor<OtapPdata> for RetryProcessor {
         effect_handler: &mut EffectHandler<OtapPdata>,
     ) -> Result<(), Error> {
         match msg {
-            Message::PData(data) => {
-                // Clone only if we need to add to retry queue AND send downstream
-                // Check if queue is full first to avoid unnecessary clone
-                if self.pending_messages.len() >= self.config.max_pending_messages {
-                    let error_msg = format!(
-                        "Retry queue is full (capacity: {}), cannot add message",
-                        self.config.max_pending_messages
-                    );
-                    log::warn!("{error_msg}");
-                    // Send NACK upstream to signal backpressure instead of forwarding message
-                    // Note: This would need to be implemented in the effect handler
-                    // For now, we'll just log and drop the message
-                    return Err(Error::ProcessorError {
-                        processor: effect_handler.processor_id(),
-                        error: error_msg,
-                    });
-                } else {
-                    // Queue has space, add message for retry and send downstream
-                    let id = self.next_message_id;
-                    self.next_message_id += 1;
-
-                    let pending = PendingMessage {
-                        data: data.clone(), // Only clone when we know we need to store AND send
-                        retry_count: 0,
-                        next_retry_time: Instant::now(),
-                        last_error: String::new(),
-                    };
-
-                    let _previous = self.pending_messages.insert(id, pending);
-                    log::debug!("Added message {id} to retry queue");
-
-                    effect_handler.send_message(data).await?;
+            Message::PData(mut data) => {
+                match data.push_reply_to(
+                    Interest::NackOnly,
+                    effect_handler.processor_id().index(),
+                    ContextData::new(0),
+                ) {
```

Comment:
To better hide the details and avoid errors, I suggest reversing the way things are done here.

```rust
effect_handler.subscribe_to(Interests::Nack, &mut data)
```